### ISSUE TITLE
`symbolic`: Improve the semantics of pointer comparisons

### ISSUE THREAD

**Issue by langston-barrett** [GaloisInc/macaw#392]
Consider `doPtrLt`:

https://github.com/GaloisInc/macaw/blob/645754178a034c22e97aac49a962c99b387aaa01/symbolic/src/Data/Macaw/Symbolic/MemOps.hs#L858-L870

First of all, there's some redundancy here: `andPred sym okP1 okP2` is strictly stronger than `both_ptrs`, which it is `and`ed with.

More concerningly, this function invents a fresh predicate (via `mkUndefinedBool`) that encodes the result of the comparison in the case that they aren't pointers to the same allocation. This means that multiple calls to `doPtrLt` with the same pointers return uncorrelated fresh predicates, so that the following pseudo-code program could report an assertion failure, even though the assertion should be unreachable:
```c
void f(void *p, void *q) {
    if (p < q) {
        if (q < p) {
            assert(false);
        }
    }
}
```
(I haven't tested this, but it should be easy to develop such a test case when we have a CLI #390). `doPtrLeq` suffers from the same problem.

It's not clear if there's really a great solution here. In the ideal world, Macaw would explore *all* and *only* the feasible executions of a program, but for these operations it appears that there's a trade-off:

- If we just *assert* (as a safety condition) that pointer comparisons only happen between pointers to the same allocation (or between raw bitvectors), we'll probably miss out on being able to simulate real code that appears in the wild.
- However, the current code is permissive enough that it explores *infeasible* executions.

Perhaps it's worth exposing more knobs with which to tune Macaw's behavior, as appropriate for downstream clients? For clients doing verification, this more permissive behavior might be fine: if you can prove the verification conditions for all the feasible executions *plus* some infeasible ones, you've still proved them for the feasible ones. However, for bug-hunting, we might might care less about missing some feasible paths, prefering fewer false positives (reports of possible bugs that cannot occur in real executions).

**Comment by RyanGlScott** [comment]
I agree that this behavior is not desirable in many applications, and it would be nice to offer easy-to-use alternatives. We've exposed a similar knob on the `crucible-llvm` side for configuring the behavior of how "undefined" memory from an uninitialized memory load should work (see [here](https://github.com/GaloisInc/crucible/blob/19e9a5243e73f639baec061faf0a243b03401d69/crucible-llvm/src/Lang/Crucible/LLVM/MemModel/Options.hs#L80-L107)).
